### ISSUE TITLE
stuck in customizing node visuals

### ISSUE THREAD

**Issue by ChrisDongWooKim** [Wouterdek/NodeNetwork#36]
I'm trying to follow your tutorial. but I'm stuck in step : customizing node visuals. Actually, I was stuck in where declare views in CustomNodeView.xaml. I added below codes into usercontrol so that I solved it though.

`xmlns:views="clr-namespace:NodeNetwork.Views;assembly=NodeNetwork"`

Now I'm stuck in below parts

<pre><code>
public partial class CustomNodeView : IViewFor<\CustomNodeViewModel\>
{
    #region ViewModel
    ...
    #endregion

    public CustomNodeView()
    {
        InitializeComponent();

        this.WhenActivated(d =>
        {
            this.WhenAnyValue(v => v.ViewModel).BindTo(this, v => v.NodeView.ViewModel).DisposeWith(d);
        });
    }
}
</code></pre>

I can not find CustomNodeViewModel. What should be in #region ViewModel? I guess both are related, right? Plz, could you give me codes or guide for me?

thx in advance.

**Comment by ChrisDongWooKim** [comment]
> I'm trying to follow your tutorial. but I'm stuck in step : customizing node visuals. Actually, I was stuck in where declare views in CustomNodeView.xaml. I added below codes into usercontrol so that I solved it though.
> 
> `xmlns:views="clr-namespace:NodeNetwork.Views;assembly=NodeNetwork"`
> 
> Now I'm stuck in below parts
> 
> ```
> 
> public partial class CustomNodeView : IViewFor<\CustomNodeViewModel\>
> {
>     #region ViewModel
>     ...
>     #endregion
> 
>     public CustomNodeView()
>     {
>         InitializeComponent();
> 
>         this.WhenActivated(d =>
>         {
>             this.WhenAnyValue(v => v.ViewModel).BindTo(this, v => v.NodeView.ViewModel).DisposeWith(d);
>         });
>     }
> }
> ```
> 
> I can not find CustomNodeViewModel. What should be in #region ViewModel? I guess both are related, right? Plz, could you give me codes or guide for me?
> 
> thx in advance.

there are backslashes between begin and end of the "CustomNodeViewModel" word. plz ignore it.

**Comment by Wouterdek** [comment]
`CustomNodeViewModel` can be any viewmodel class that inherits from NodeViewModel. See [the previous chapter](https://wouterdek.github.io/NodeNetwork/cookbook/custom_nodes) for details on how you can create such a class. In the specific example of the previous chapter, this would be `HelloWorldNode`.

> What should be in #region ViewModel?

This is boilerplate code to create a reactive property called ViewModel. I have modified the chapter to explain this.

    public static readonly DependencyProperty ViewModelProperty = 
        DependencyProperty.Register(nameof(ViewModel), typeof(CustomViewModel), typeof(CustomView), new PropertyMetadata(null));

    public CustomViewModel ViewModel
    {
        get => (CustomViewModel)GetValue(ViewModelProperty);
        set => SetValue(ViewModelProperty, value);
    }

    object IViewFor.ViewModel
    {
        get => ViewModel;
        set => ViewModel = (CustomViewModel)value;
    }

**Comment by ChrisDongWooKim** [comment]
Thx to answer me, Wouterdek.
I got some more questions about your answer. I have tried to follow your answer like below codes. there were many errors came out though.

```{.cs}
    public partial class CustomNodeView : IViewFor<CustomNodeViewModel>
    {

        public CustomNodeView()
        {
            InitializeComponent();

            this.WhenActivated(d =>
            {
                this.WhenAnyValue(v => v.ViewModel).BindTo(this, v => v.NodeView.ViewModel).DisposeWith(d);
            });
        }

        public static readonly DependencyProperty ViewModelProperty =
        DependencyProperty.Register(nameof(ViewModel), typeof(CustomNodeView), typeof(CustomNodeViewModel), new PropertyMetadata(null));


        public CustomNodeViewModel ViewModel
        {
            get => (CustomNodeViewModel)GetValue(ViewModelProperty);
            set => SetValue(ViewModelProperty, value);
        }

        object IViewFor.ViewModel
        {
            get => ViewModel;
            set => ViewModel = (CustomNodeViewModel)value;
        }
    }

    public class CustomNodeViewModel : NodeViewModel
    {
        public ValueNodeInputViewModel<string> NameInput { get; }
        public ValueNodeOutputViewModel<string> TextOutput { get; }

    }

both of view model have a lot of errors that I can not solve them. if you have completed codes of this parts, could you plz share it with us? so that I can recognize what I did wrong. To be honest, it will be very helpful for us, if every parts have a sample code in cookbook tutorial.

thx in advance.

**Comment by Wouterdek** [comment]
The codegen example application includes a complete example of how to create a custom node view, maybe that clears things up a bit?
- [View XAML](https://github.com/Wouterdek/NodeNetwork/blob/master/ExampleCodeGenApp/Views/CodeGenNodeView.xaml)
- [View codebehind](https://github.com/Wouterdek/NodeNetwork/blob/master/ExampleCodeGenApp/Views/CodeGenNodeView.xaml.cs)
- [ViewModel](https://github.com/Wouterdek/NodeNetwork/blob/master/ExampleCodeGenApp/ViewModels/CodeGenNodeViewModel.cs)

**Comment by ChrisDongWooKim** [comment]
Thx, Wouterdek.
I have done what you said before but didn't work and occured many errors in viewmodel. So I asked you to get sample sources to solve this errors. BTW, I opened tutorial source using VS 2019 today. then, all errors have gone. I have used VS 2015 till yesterday. I guess latest version of VS can support latest .net framework.
Also, the links you posted are also so helpful for me to understand tutorial source.

have a nice day. Wouterdek :D
